<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="AG Schissler" />


<title>Week 9 / Exploratory Data Analysis, ggplot2, Lab 07</title>

<style type="text/css">
  
  /* Whole document - standard notebook size and margins */
  body {
      font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
      max-width: 8.5in;
      margin: auto;
      background: #FFFFFF;
      padding: 1in 1in 1in 1in;
  }
  
</style>





<link href="data:text/css;charset=utf-8,%3Cstyle%20type%3D%22text%2Fcss%22%3E%0A%0A%0Abody%20%7B%0Afont%2Dfamily%3A%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Amax%2Dwidth%3A%208%2E5in%3B%0Amargin%3A%20auto%3B%0Abackground%3A%20%23FFFFFF%3B%0Apadding%3A%201in%201in%201in%201in%3B%0A%7D%0A%0Ah1%20%7B%0Afont%2Dsize%3A%2050px%3B%0Acolor%3A%20%23000000%3B%0Atext%2Dalign%3Acenter%3B%0Amargin%2Dbottom%3A15px%3B%0A%7D%0Ah1%2Bp%20%7B%0A%0Acolor%3A%20%23000000%3B%0Afont%2Dsize%3A%2020px%3B%0Aline%2Dheight%3A%202em%3B%0Amargin%2Dtop%3A%200%2E5em%3B%0Amargin%2Dbottom%3A%200em%20%21important%3B%0Atext%2Dalign%3A%20center%3B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0A%0Ah2%20%7B%0Acolor%3A%20%23000000%3B%0A%7D%0A%0Adt%20%7B%0Afloat%3A%20left%3B%0Aclear%3A%20left%3B%0Awidth%3A%2017%25%3B%0A%0A%7D%0Add%20%7B%0Amargin%2Dleft%3A%2017%25%3B%0A%7D%0Ap%20%7B%0Amargin%2Dtop%3A0%3B%0Amargin%2Dbottom%3A7px%3B%0A%7D%0A%0Ablockquote%20%7B%0Atext%2Dalign%3A%20left%3B%20%7D%0A%0Aa%20%7B%0Atext%2Ddecoration%3A%20none%3B%0Acolor%3A%20%232980d6%3B%0A%7D%0Aa%3Ahover%2C%20a%3Aactive%20%7B%0Abackground%2Dcolor%3A%20%232980d6%3B%0Acolor%3A%20%23FFFFFF%3B%0Atext%2Ddecoration%3A%20none%3B%0Atext%2Dshadow%3A%201px%201px%201px%20%23333%3B%0A%7D%0A%0Ahr%20%7B%0Acolor%3A%20%23A6A6A6%3B%0A%7D%0Atable%20%7B%0Awidth%3A%20100%25%3B%0Aborder%2Dcollapse%3Acollapse%3B%0Aborder%2Dspacing%3A0%3B%0Adisplay%3Atable%0A%7D%0Atable%2EminimalistBlack%20%7B%0Aborder%3A%203px%20solid%20%23000000%3B%0Awidth%3A%20100%25%3B%0Atext%2Dalign%3A%20left%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%2EminimalistBlack%20td%2C%20table%2EminimalistBlack%20th%20%7B%0Apadding%3A%205px%204px%3B%0A%7D%0Atable%2EminimalistBlack%20tbody%20td%20%7B%0Afont%2Dsize%3A%2015px%3B%0A%7D%0Atable%2EminimalistBlack%20thead%20%7B%0Abackground%3A%20%23CFCFCF%3B%0Abackground%3A%20%2Dmoz%2Dlinear%2Dgradient%28top%2C%20%23dbdbdb%200%25%2C%20%23d3d3d3%2066%25%2C%20%23CFCFCF%20100%25%29%3B%0Abackground%3A%20%2Dwebkit%2Dlinear%2Dgradient%28top%2C%20%23dbdbdb%200%25%2C%20%23d3d3d3%2066%25%2C%20%23CFCFCF%20100%25%29%3B%0Abackground%3A%20linear%2Dgradient%28to%20bottom%2C%20%23dbdbdb%200%25%2C%20%23d3d3d3%2066%25%2C%20%23CFCFCF%20100%25%29%3B%0Aborder%2Dbottom%3A%203px%20solid%20%23000000%3B%0A%7D%0Atable%2EminimalistBlack%20thead%20th%20%7B%0Afont%2Dsize%3A%2015px%3B%0Afont%2Dweight%3A%20bold%3B%0Acolor%3A%20%23000000%3B%0Atext%2Dalign%3A%20left%3B%0A%7D%0Atable%2EminimalistBlack%20tfoot%20%7B%0Afont%2Dsize%3A%2014px%3B%0Afont%2Dweight%3A%20bold%3B%0Acolor%3A%20%23000000%3B%0Aborder%2Dtop%3A%203px%20solid%20%23000000%3B%0A%7D%0Atable%2EminimalistBlack%20tfoot%20td%20%7B%0Afont%2Dsize%3A%2014px%3B%0A%7D%0A%0A%2Eanswer%5Fdiv%20%7B%0Awidth%3A100%25%3B%0Aheight%3A2in%3B%0Afloat%3Aleft%3B%0A%7D%0A%2Elong%5Fanswer%5Fdiv%7B%0Awidth%3A100%25%3B%0Aheight%3A3in%3B%0Afloat%3Aleft%3B%0A%7D%0A%0A%40media%20print%20%7B%0A%2Epagebreak%20%7B%20page%2Dbreak%2Dbefore%3A%20always%3B%20padding%2Dtop%3A1in%3B%0A%7D%20%7D%0Adiv%2Esolutions%20%2A%20%7B%0Acolor%3A%23000099%3B%0A%7D%0A%3C%2Fstyle%3E%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Week 9 / Exploratory Data Analysis, ggplot2, Lab 07</h1>
<h4 class="author">AG Schissler</h4>
<h4 class="date">Wednesday October 30, 2019</h4>



<div id="i.-opening" class="section level2">
<h2>I. Opening</h2>
<div id="i.-start-of-class-work" class="section level3">
<h3>I. Start-of-class work</h3>
<ol style="list-style-type: decimal">
<li>Log in the your workstation.</li>
<li>Open the RStudio application on your machine.</li>
<li>Open <a href="https://github.com/grizant/STAT445/blob/master/meeting_agendas/19_ggplot2_continueLab07.Rmd">19_ggplot2_continueLab07.Rmd</a> from website so that you can work along the lecture.</li>
<li>Open <a href="https://github.com/grizant/STAT445/blob/master/labs/lab_07_ggplot2_graphics.Rmd">lab_07_ggplot2_graphics.Rmd</a> to have the lab available.</li>
</ol>
</div>
<div id="this-weeks-agenda" class="section level3">
<h3>This week’s agenda</h3>
<ul>
<li>Continue <em>Working with data</em> Unit II.</li>
<li>Focus on using R to plot data to conduct exploratory data analysis (EDA).</li>
<li>Now <code>ggplot2</code> will be used.</li>
</ul>
</div>
<div id="today-learning-outcomes" class="section level3">
<h3>Today learning outcomes</h3>
<p>Students will be able to</p>
<ul>
<li>use <code>ggplot2</code> to plot data</li>
</ul>
</div>
</div>
<div id="ii.-a-detailed-example-a-custom-plot-with-the-r-package-nof1" class="section level2">
<h2>II. A detailed example a custom plot with the R package <code>nof1</code></h2>
<pre class="r"><code>#' Rose Plot Color Selection
#' 
#' \code{select_pathway_colors} provides an interface to select the colors corresponding to pathways for a given direction within a rose plot.
#'
#' @details Rose plots are designed to visually inspect differential pathway expression scaled by statistical significance for rapid interpretation of N-of-1-\emph{pathways} results. As such, only a few pathways are easy to see at a time. This function will automatically select colors when at most 9 pathways are plotted (corresponding to 18 'petals', one for each direction). For more pathways, the user must supply the colors.
#'
#' @param num_pathways a numeric value indicating the number pathways to plot, values 1 to 9 for automatic coloring
#' @param colors_up a character vector with the colors for the higher expressed pathways
#' @param colors_down a character vector with the colors for the lower expressed pathways
#' @param gradient_spread a numeric value indicating the gradient of colors during automatically selection. See \code{\link[RColorBrewer]{brewer.pal}} for details.
#'
#' @return a list of length 2 each element containing a character vector color codes (hexidecimal representation) for the pathways in both up (above horizonal axis) and down (below horizonal axis) directions.
#'
#' @note The default color scheme is intended convey the appearance of blue sky and brown earth below. Enjoy.
#'
#' @references Schissler, A. Grant, et al. &quot;Analysis of aggregated cell-cell statistical distances within pathways unveils therapeutic-resistance mechanisms in circulating tumor cells.&quot; Bioinformatics 32.12 (2016): i80-i89.
#'
#' @seealso \code{link{create_rose_plot}}, \code{link{create_multi_rose_plot}}
#' 
#' @examples
#' ## Show simple automatic example
#' select_pathway_colors(5)
#' ## Change gradient to be more subtle
#' select_pathway_colors(5, gradient_spread = 5)
#' ## select custom pathways
#' select_pathway_colors(num_pathways = 3,
#'                       colors_up = c(&quot;#C6DBEF&quot;, &quot;#9ECAE1&quot;, &quot;#6BAED6&quot;),
#'                       colors_down = c(&quot;#FEE391&quot;, &quot;#FEC44F&quot;, &quot;#FE9929&quot;))
#' 
#' @export

select_pathway_colors &lt;- function(num_pathways, colors_up = NULL, colors_down = NULL, gradient_spread = 9) {
    if (is.null(colors_up) &amp; is.null(colors_down)) {  
        if (!requireNamespace(&quot;RColorBrewer&quot;, quietly = TRUE)) {
            stop(&quot;RColorBrewer needed to automatically choose rose plot colors. Please install it or specify your own color vectors.&quot;, call. = FALSE)
        } else {
            ## check to make sure gradient spread can cover the pathways
            if (num_pathways &lt;= gradient_spread) {
                ## use color brewer to select the colors
                ## select the middle values in the gradient
                middle_index &lt;- ceiling(gradient_spread/2)
                lower_index &lt;- middle_index - floor(num_pathways/2)
                upper_index &lt;- middle_index + floor(num_pathways/2)
                if (num_pathways %% 2 == 0) lower_index &lt;- lower_index + 1
                ## index the blue and earth tones
                colors_up &lt;- RColorBrewer::brewer.pal(gradient_spread,
                                                      &quot;Blues&quot;)[lower_index:upper_index]
                colors_down &lt;- RColorBrewer::brewer.pal(gradient_spread,
                                                        &quot;YlOrBr&quot;)[lower_index:upper_index]
                return(list(colors_up = colors_up, colors_down = colors_down))
            } else stop(&quot;Automatic color selection does not work with more than pathways than gradient selected. Please specify your own colors.&quot;)
        }
    } else {
        if ((length(colors_up) == num_pathways) &amp; (length(colors_down) == num_pathways)) {
            ## simply use the user-specified colors
            return(list(colors_up = colors_up, colors_down = colors_down))
        } else {
            stop(&quot;Color vectors must match the number of pathways in the rose plot.&quot;)
        }
    }
}

#' Rose Plot Construction
#'
#' \code{create_rose_plot} constructs a rose plot visualization of pathway differential expression derived from an application of N-of-1-\emph{pathways}.
#'
#' @details This function produces a rose plot displaying a profile pathway differential expression metrics (currently only log P-values are plotted). A rose plot is in this setting is a bar plot in polar coordinates with positive scores above the horizon and negative below. Therefore, the bars become the 'petals' in the rose plot.
#' 
#' @param pathway_scores a data frame of pathway scores from \code{\link{compute_nof1_pathways}}
#' @param which_pathways a character vector of path_id or numeric vector indexing the pathways to plot
#' @param plot_values a character value specifying which metric to plot. For now, the only option is the log_pvalue. Later versions will implement z-scores and pathway scores.
#' @param use_desc a logical indicating to label pathways by the description rather than the ID
#' @param colors_up a character vector with the colors for the higher expressed pathways
#' @param colors_down a character vector with the colors for the lower expressed pathways
#' @param legend a logical whether to display a legend
#' @param ... other arguments for \code{\link{select_pathway_colors}}
#'
#' @return a ggplot2 object containing a rose plot visualization
#' 
#' @references
#' Gardeux, Vincent, et al. &quot;'N-of-1-\emph{pathways}' unveils personal deregulated mechanisms from a single pair of RNA-Seq samples: towards precision medicine.&quot; Journal of the American Medical Informatics Association 21.6 (2014): 1015-1025.
#' 
#' Schissler, A. Grant, et al. &quot;Dynamic changes of RNA-sequencing expression for precision medicine: N-of-1-\emph{pathways} Mahalanobis distance within pathways of single subjects predicts breast cancer survival.&quot; Bioinformatics 31.12 (2015): i293-i302.
#' 
#' Schaefer, Carl F., et al. &quot;PID: the pathway interaction database.&quot; Nucleic acids research 37.suppl 1 (2009): D674-D679.
#'
#' Weinstein, John N., et al. &quot;The cancer genome atlas pan-cancer analysis project.&quot; Nature genetics 45.10 (2013): 1113-1120.
#'
#' Schissler, A. Grant, et al. &quot;Analysis of aggregated cell-cell statistical distances within pathways unveils therapeutic-resistance mechanisms in circulating tumor cells.&quot; Bioinformatics 32.12 (2016): i80-i89.
#'
#' @seealso \code{\link{normalize_gene_data}} \code{\link{compute_multi_nof1_pathways}} \code{link{pid_filtered}} \code{link{BRCA_TCGA_BH_A1EV}} \code{link{pid_desc}}
#'
#' @examples
#' \dontrun{
#' ## Here a rose plot is produced for pre-computed N-of-1-pathways MD scores
#' ## derived from a single the cancer genome atlas (TCGA) breast cancer patient.
#' ## and pathways (gene sets) defined by the Pathway Interaction Database (PID).
#' data(A7_A0DB_PID_MD_scores)
#' one_scores &lt;- A7_A0DB_PID_MD_scores
#' ## plot the first five most significant pathways
#' create_rose_plot(one_scores, 1:5)
#' ## select pathways of interest, only some of which are actually scored
#' my_pathways &lt;- c(&quot;200028&quot;, &quot;200035&quot;, &quot;200097&quot;, &quot;200145&quot;, &quot;200192&quot;)
#' create_rose_plot(one_scores, my_pathways)
#' ## using indeces
#' create_rose_plot(one_scores, seq(20, nrow(one_scores), length.out = 9))
#' }
#' 
#' @export

create_rose_plot &lt;- function(pathway_scores, which_pathways, plot_values = &quot;log_pvalue&quot;, use_desc = T, colors_up = NULL, colors_down = NULL, legend = T,...) {
    ## 1. Start by processing the pathway_scores data frame and selecting colors
    ## 1.1 select only the pathways of interest and values of interest
    ## value_name &lt;- ifelse(!use_zscore, &quot;pathway_score&quot;, &quot;zscore&quot;)
    ## need to add other options in rose plot
    value_name &lt;- &quot;p_value&quot;
    ## check to see if all pathways indicated are present
    if (class(which_pathways) == &quot;character&quot;) {
         if (sum(rownames(pathway_scores) %in% which_pathways) &lt; length(which_pathways)) {
             warning(&quot;Not all pathways were found, making a partial rose plot.&quot;)
             which_pathways &lt;- rownames(pathway_scores)[rownames(pathway_scores) %in% which_pathways]
         }
    }
    rose_scores &lt;- pathway_scores[which_pathways, c(value_name, &quot;direction&quot;)]
    names(rose_scores) &lt;- c(&quot;value&quot;, &quot;direction&quot;)
    ## transform the pvalue
    rose_scores$value &lt;- -log( rose_scores$value, base = exp(1) )
    ## check if no pathways are selected
    num_pathways &lt;- nrow(rose_scores)
    if (num_pathways == 0) stop(&quot;Rose plot cannot be created. No valid pathways selected.&quot;)
    ## make pathway ID a factor for plotting
    ## rose_scores$ID &lt;- factor(rownames(rose_scores))
    ## 1.2 select the colors
    ## rose_colors &lt;- select_pathway_colors(length(which_pathways), colors_up = colors_up, colors_down = colors_down)
    rose_colors &lt;- select_pathway_colors(num_pathways, colors_up = colors_up, colors_down = colors_down, ...)
    ## 2. construct data structure for plotting
    ## 2.1 construct the bins for the position of the bars
    pos &lt;- seq(0, 2*num_pathways - 1, by = 1)
    rose_data &lt;- data.frame(factor(1:(2*num_pathways)), 0, pos)
    names(rose_data) &lt;- c(&quot;pathway&quot;, &quot;value&quot;, &quot;pos&quot;)
    ## choose positions based on desired location up or down and make values positive
    for (j in 1:nrow(rose_data)) {
        ifelse( rose_scores$direction[j] &gt; 0,
               rose_data$value[j] &lt;- rose_scores$value[j],
               rose_data$value[j + num_pathways] &lt;- rose_scores$value[j])
    }
    ## 2.2 transform values (radii) to maintain equal area instead of radii
    rose_data$value &lt;- sqrt(rose_data$value)
    ## 2.3 make pathway a factor and optionally add description
    rose_labels &lt;- rownames(rose_scores)
    if (use_desc) rose_labels &lt;- pathway_scores[which_pathways, &quot;pathway_desc&quot;]
    rose_data$pathway &lt;- factor( rose_data$pathway, labels = c(paste0(rose_labels,&quot; +&quot;),paste0(rose_labels,&quot; -&quot;) ))

    ## make pathway a factor for plotting
    rose_data$pathway &lt;- factor(rose_data$pathway)
    ## 3. create ggplot object
    p0 &lt;- ggplot2::ggplot(rose_data, ggplot2::aes_string(fill = &quot;pathway&quot;, x = &quot;pos&quot;, y = &quot;value&quot;))
    ## create bar plot to later convert to polar coordinates
    width &lt;- 1
    p1 &lt;- p0 + ggplot2::geom_bar(width = width, stat = &quot;identity&quot;)
    ## create rose and shift pathways to start from horizontal axis
    p2 &lt;- p1 + ggplot2::coord_polar(start = -pi*(nrow(rose_data)) - pi/2)
    ## update y scales to show breaks of interest
    ## p3 &lt;- p2 + scale_x_continuous(labels = NULL, breaks = pos + width/2, minor_breaks = NULL)
    p3 &lt;- p2 + ggplot2::scale_x_continuous(labels = NULL, breaks = (pos + width/2)[c(num_pathways,num_pathways*2)], minor_breaks = NULL)
    ## customize fill scale
    p4 &lt;- p3 + ggplot2::scale_fill_manual(values = unname(unlist(rose_colors)), guide = ggplot2::guide_legend(title = &quot;Pathways&quot;))
    ## customize theme
    ## p5 &lt;- p4 + theme_bw() + xlab(&quot;&quot;)
    p5 &lt;- p4 + ggplot2::xlab(&quot;&quot;)
    ## add darker horizontal lines to guide eye for up vs. down regulation
    p6 &lt;- p5 + ggplot2::theme(panel.grid.major.x = ggplot2::element_line(colour = &quot;black&quot;, linetype = 1))
    ## update y scales to show breaks of interest (significance levels)
    y_labels &lt;- c(0.05, 0.01, 0.001)
    ybreaks &lt;- sqrt (-log(y_labels, exp(1)) )
    ## p7 &lt;- p6 + ggplot2::scale_y_continuous(labels = NULL, breaks = ybreaks, minor_breaks = NULL) +
       ## theme(panel.grid.major.y = ggplot2::element_line(colour = &quot;black&quot;, linetype = 1))
    p7 &lt;- p6 + ggplot2::scale_y_continuous(labels = as.character(y_labels),
                                  breaks = ybreaks, minor_breaks = NULL) +
        ggplot2::theme(panel.grid.major.y = ggplot2::element_line(colour = &quot;black&quot;, linetype = 1))
    ## update y scales labeling
    ## p8 &lt;- p7 + theme(axis.ticks = element_blank(), axis.text.y = element_blank(), axis.title.y = element_blank())
    p8 &lt;- p7 + ggplot2::labs(y = &quot;p-value&quot;)
    ## remove dark color from background
    p9 &lt;- p8 + ggplot2::theme(strip.background = ggplot2::element_rect(fill = 'white'))
    ## move legend 
    if (legend) {
        p10 &lt;- p9 + ggplot2::theme(legend.position = &quot;right&quot;, legend.direction = &quot;vertical&quot;)
    } else p10 &lt;- p9 + ggplot2::theme(legend.position = &quot;none&quot;)
    p10
}

#' Multiple Rose Plots Construction
#'
#' \code{create_multi_rose_plot} constructs multiple rose plots to compare profiles of pathway differential expression derived from applying N-of-1-\emph{pathways}.
#'
#' @details This function produces side-by-side rose plots displaying aprofile pathway differential expression metrics (currently only log P-values are plotted). A rose plot is in this setting is a bar plot in polar coordinates with positive scores above the horizon and negative below. Therefore, the bars become the 'petals' in the rose plot.
#' 
#' @param pathway_scores a list of data frames containing pathway scores obtained from \code{\link{compute_multi_nof1_pathways}}
#' @param which_pathways a character vector of path_id or numeric vector indexing the pathways to plot
#' @param plot_values a character value specifying which metric to plot. For now, the only option is the log_pvalue. Later versions will implement z-scores and pathway scores.
#' @param use_desc a logical indicating to label pathways by the description rather than the ID
#' @param colors_up a character vector with the colors for the higher expressed pathways
#' @param colors_down a character vector with the colors for the lower expressed pathways
#' @param legend a logical whether to display a legend
#' @param ... other arguments for \code{\link{select_pathway_colors}}
#'
#' @return a ggplot2 object containing a rose plot visualization
#'
#' @note The selected pathways must be present for each item in the pathway scores list.
#' 
#' @references
#' Gardeux, Vincent, et al. &quot;'N-of-1-\emph{pathways}' unveils personal deregulated mechanisms from a single pair of RNA-Seq samples: towards precision medicine.&quot; Journal of the American Medical Informatics Association 21.6 (2014): 1015-1025.
#' 
#' Schissler, A. Grant, et al. &quot;Dynamic changes of RNA-sequencing expression for precision medicine: N-of-1-\emph{pathways} Mahalanobis distance within pathways of single subjects predicts breast cancer survival.&quot; Bioinformatics 31.12 (2015): i293-i302.
#' 
#' Schaefer, Carl F., et al. &quot;PID: the pathway interaction database.&quot; Nucleic acids research 37.suppl 1 (2009): D674-D679.
#'
#' Weinstein, John N., et al. &quot;The cancer genome atlas pan-cancer analysis project.&quot; Nature genetics 45.10 (2013): 1113-1120.
#'
#' Schissler, A. Grant, et al. &quot;Analysis of aggregated cell-cell statistical distances within pathways unveils therapeutic-resistance mechanisms in circulating tumor cells.&quot; Bioinformatics 32.12 (2016): i80-i89.
#'
#' @seealso \code{\link{normalize_gene_data}} \code{\link{compute_multi_nof1_pathways}} \code{link{pid_filtered}} \code{link{BRCA_TCGA_BH_A1EV}} \code{link{pid_desc}}
#'
#' @examples
#' \dontrun{
#' ## Here a rose plot is produced for pre-computed N-of-1-pathways MD scores
#' ## derived from a single the cancer genome atlas (TCGA) breast cancer patient.
#' ## and pathways (gene sets) defined by the Pathway Interaction Database (PID).
#' data(BRCA_TCGA_4_PID_scores)
#' multi_scores &lt;- BRCA_TCGA_4_PID_scores
#' ## trim samples names to make more visible on plot
#' names(multi_scores) &lt;- gsub(&quot;TCGA\\.&quot;,&quot;&quot;, names(multi_scores))
#' ## select pathways of interest
#' my_pathways &lt;- c(&quot;200177&quot;, &quot;200190&quot;, &quot;200026&quot;, &quot;200041&quot;,
#' &quot;200224&quot;, &quot;200090&quot;,&quot;200096&quot;, &quot;200123&quot;, &quot;200091&quot;)
#' ## make plots
#' create_multi_rose_plot(multi_scores, my_pathways)
#' }
#' 
#' @export

create_multi_rose_plot &lt;- function(pathway_scores, which_pathways, plot_values = &quot;log_pvalue&quot;, use_desc = T, colors_up = NULL, colors_down = NULL, legend = T,...) {
    ## 1. Start by processing the pathway_scores data frame and selecting colors
    ## 1.1 select only the pathways of interest and values of interest
    ## value_name &lt;- ifelse(!use_zscore, &quot;pathway_score&quot;, &quot;zscore&quot;)
    ## need to add other options in rose plot
    value_name &lt;- &quot;p_value&quot;
    ## check to see if all pathways indicated are present
    if (class(which_pathways) == &quot;character&quot;) {
        ## get counts of selected pathways
        pathway_counts &lt;- unlist(lapply(pathway_scores, function(tmp_scores){
            sum(rownames(tmp_scores) %in% which_pathways)
        }))
         if (any(pathway_counts &lt; length(which_pathways))) {
             stop(&quot;Not all pathways were found. Please ensure all pathways selected are scored for each element of the pathway_scores list.&quot;)
         }
    }
    ##
    num_pathways &lt;- length(which_pathways)
    ## 1.2 select the colors
    ## rose_colors &lt;- select_pathway_colors(length(which_pathways), colors_up = colors_up, colors_down = colors_down)
    rose_colors &lt;- select_pathway_colors(num_pathways, colors_up = colors_up, colors_down = colors_down,...)
    ## 1.3 create the rose plot data
    ## tmp_scores &lt;- pathway_scores[[1]]
    rose_data &lt;- lapply(pathway_scores, function(tmp_scores) {
        tmp_scores &lt;- tmp_scores[which_pathways, c(value_name, &quot;direction&quot;, &quot;pathway_desc&quot;)]
        names(tmp_scores) &lt;- c(&quot;value&quot;, &quot;direction&quot;, &quot;pathway_desc&quot;)
        ## transform the pvalue
        tmp_scores$value &lt;- -log( tmp_scores$value, base = exp(1) )
        ## 2. construct data structure for plotting
        ## 2.1 construct the bins for the position of the bars
        pos &lt;- seq(0, 2*num_pathways - 1, by = 1)
        tmp_data &lt;- data.frame(factor(1:(2*num_pathways)), 0, pos)
        names(tmp_data) &lt;- c(&quot;pathway&quot;, &quot;value&quot;, &quot;pos&quot;)
        ## choose positions based on desired location up or down and make values positive
        for (j in 1:nrow(tmp_data)) {
            ifelse( tmp_scores$direction[j] &gt; 0,
                   tmp_data$value[j] &lt;- tmp_scores$value[j],
                   tmp_data$value[j + num_pathways] &lt;- tmp_scores$value[j])
        }
        ## 2.2 transform values (radii) to maintain equal area instead of radii
        tmp_data$value &lt;- sqrt(tmp_data$value)
        ## 2.3 make pathway a factor and optionally add description
        rose_labels &lt;- rownames(tmp_scores)
        if (use_desc) rose_labels &lt;- tmp_scores[which_pathways, &quot;pathway_desc&quot;]
        tmp_data$pathway &lt;- factor( tmp_data$pathway, labels = c(paste0(rose_labels,&quot; +&quot;),paste0(rose_labels,&quot; -&quot;) ))
        tmp_data
    })
    
    ## 2.4 annotate with pair label
    ## retrieve pair label
    num_pathways &lt;- nrow(rose_data[[1]])/2
    pair_names &lt;- rep(names(rose_data), each = 2*num_pathways)
    ## convert to data frame
    rose_data &lt;- do.call(&quot;rbind&quot;, rose_data)
    rose_data$pair &lt;- pair_names

    ## 2.5 automatically order the plots
    ## tmp_data &lt;- split(rose_data, rose_data$pair)[[1]]
    order_data &lt;- do.call(&quot;rbind&quot;, lapply(split(rose_data, rose_data$pair), function(tmp_data) {
        average_value &lt;- 1/mean(c(tmp_data[1:num_pathways,&quot;value&quot;], -1*tmp_data[(num_pathways+1):nrow(tmp_data),&quot;value&quot;]))
        count_down &lt;- sum(tmp_data[(num_pathways+1):nrow(tmp_data),&quot;value&quot;] != 0)
        data.frame(average_value, count_down)
    }))
    ## order by count down then average
    order_data &lt;- order_data[order(order_data[,&quot;count_down&quot;],order_data[,&quot;average_value&quot;]),]
    ## add pair label
    rose_data$pair &lt;- factor(pair_names, levels = row.names(order_data))

    ## 3. create ggplot object
    ## 3.1 retrieve plotting characteristics
    ## retrieve the position sequence
    pos &lt;- unique(rose_data$pos)

    ## make pathway a factor
    rose_data$pathway &lt;- factor(rose_data$pathway)
    ## 3.2 build up the object in layers
    p0 &lt;- ggplot2::ggplot(rose_data, ggplot2::aes_string(fill = &quot;pathway&quot;, x = &quot;pos&quot;, y = &quot;value&quot;))
    ## create bar plot to later convert to polar coordinates
    width &lt;- 1
    p1 &lt;- p0 + ggplot2::geom_bar(width = width, stat = &quot;identity&quot;)
    ## create rose and shift pathways to start from horizontal axis
    p2 &lt;- p1 + ggplot2::coord_polar(start = -pi*(num_pathways*2) - pi/2)
    ## update y scales to show breaks of interest
    ## p3 &lt;- p2 + scale_x_continuous(labels = NULL, breaks = pos + width/2, minor_breaks = NULL)
    p3 &lt;- p2 + ggplot2::scale_x_continuous(labels = NULL, breaks = (pos + width/2)[c(num_pathways,num_pathways*2)], minor_breaks = NULL)
    ## customize fill scale
    p4 &lt;- p3 + ggplot2::scale_fill_manual(values = unname(unlist(rose_colors)), guide = ggplot2::guide_legend(title = &quot;Pathways&quot;))
    ## customize theme
    ## p5 &lt;- p4 + theme_bw() + ggplot2::xlab(&quot;&quot;)
    p5 &lt;- p4 + ggplot2::xlab(&quot;&quot;)
    ## add darker horizontal lines to guide eye for up vs. down regulation
    p6 &lt;- p5 + ggplot2::theme(panel.grid.major.x = ggplot2::element_line(colour = &quot;black&quot;, linetype = 1))
    ## update y scales to show breaks of interest (significance levels)
    y_labels &lt;- c(0.05, 0.01, 0.001)
    ybreaks &lt;- sqrt (-log(y_labels, exp(1)) )
    ## p7 &lt;- p6 + ggplot2::scale_y_continuous(labels = NULL, breaks = ybreaks, minor_breaks = NULL) +
       ## ggplot2::theme(panel.grid.major.y = ggplot2::element_line(colour = &quot;black&quot;, linetype = 1))
    p7 &lt;- p6 + ggplot2::scale_y_continuous(labels = as.character(y_labels),
                                  breaks = ybreaks, minor_breaks = NULL) +
        ggplot2::theme(panel.grid.major.y = ggplot2::element_line(colour = &quot;black&quot;, linetype = 1))
    ## update y scales labeling
    ## p8 &lt;- p7 + ggplot2::theme(axis.ticks = element_blank(), axis.text.y = element_blank(), axis.title.y = element_blank())
    p8 &lt;- p7 + ggplot2::labs(y = &quot;p-value&quot;)
    ## remove dark color from background
    p9 &lt;- p8 + ggplot2::theme(strip.background = ggplot2::element_rect(fill = 'white'))
    ## move legend 
    if (legend) {
        p10 &lt;- p9 + ggplot2::theme(legend.position = &quot;right&quot;, legend.direction = &quot;vertical&quot;)
    } else p10 &lt;- p9 + ggplot2::theme(legend.position = &quot;none&quot;)
    ## facet by pair
    p11 &lt;- p10 + ggplot2::facet_wrap(~ pair)
    p11
}</code></pre>
<div id="lets-see-it-in-action-from-a-vignette" class="section level3">
<h3>Let’s see it in action from a vignette</h3>
<p><a href="http://www.lussiergroup.org/publications/nof1pathways-vignette_v2.html" class="uri">http://www.lussiergroup.org/publications/nof1pathways-vignette_v2.html</a></p>
</div>
</div>
<div id="iii.-please-continue-lab-07" class="section level2">
<h2>III. Please continue Lab 07</h2>
<ul>
<li>Questions for the whole group? Please don’t be shy it could save us all effort.</li>
<li>Please begin and I will answer questions in the order I receive them and considering who I have helped previously in class.</li>
</ul>
</div>
<div id="iv.-summary" class="section level2">
<h2>IV. Summary</h2>
<p>What is easier to do in base graphics than in ggplot2? Vice versa?</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
